src/data/bucket/fill_extrusion_bucket.js

```diff
@@ -588,6 +600,29 @@ function tileToMeter(canonical: CanonicalTileID) {
     return circumferenceAtEquator * 2 * exp / (exp * exp + 1) / EXTENT / (1 << canonical.z);
 }
 
+function isAOConcaveAngle(p2, p1, p3) {
+    if (p2.x < 0 || p2.x >= EXTENT || p1.x < 0 || p1.x >= EXTENT || p3.x < 0 || p3.x >= EXTENT) {
+        return false; // angles are not processed for edges that extend over tile borders
+    }
+    const a = p3.sub(p1);
+    const an = a.perp();
+    const b = p1.sub(p2);
+    const ab = a.x * b.x + a.y * b.y;
+    const cosABSquare = ab * ab / ((a.x * a.x + a.y * a.y) * (b.x * b.x + b.y * b.y));
+    const dotProductWithNormal = an.x * b.x + an.y * b.y;
+
+    // Heuristics: don't shade concave angles above 147° (sqrt(arccos(0.7))).
+    return cosABSquare < 0.7 && dotProductWithNormal > 0;
```

Comment:
Looking at the equation `cos^2(x) < 0.7` the solution is roughly `0.579 < x < 2.57`. This means that really sharp concave angles are discarded as well (I believe).